Lib/test/test_typing.py

```diff
@@ -8675,6 +8675,33 @@ class X(Annotated[int, (1, 10)]): ...
         self.assertEqual(X.__mro__, (X, int, object),
                          "Annotated should be transparent.")
 
+    def test_annotated_cached_with_types(self):
+        class A(str): ...
+        class B(str): ...
+
+        class Message:
+            field_a: Annotated[str, A("X")]
+            field_b: Annotated[str, A("Y")]
+            field_c: Annotated[int, 1]
+        class Form:
+            box_a: Annotated[str, B("X")]
+            box_b: Annotated[str, B("Y")]
+            box_c: Annotated[int, 1.0]
+
+        hints = get_type_hints(Message, include_extras=True)
+        self.assertEqual(type(hints["field_a"].__metadata__[0]), A)
+        self.assertEqual(hints["field_a"].__metadata__[0], A("X"))
+        self.assertEqual(type(hints["field_b"].__metadata__[0]), A)
+        self.assertEqual(hints["field_b"].__metadata__[0], A("Y"))
+        self.assertEqual(type(hints["field_c"].__metadata__[0]), int)
+
+        hints = get_type_hints(Form, include_extras=True)
+        self.assertEqual(type(hints["box_a"].__metadata__[0]), B)
+        self.assertEqual(hints["box_a"].__metadata__[0], B("X"))
+        self.assertEqual(type(hints["box_b"].__metadata__[0]), B)
+        self.assertEqual(hints["box_b"].__metadata__[0], B("Y"))
+        self.assertEqual(type(hints["box_c"].__metadata__[0]), float)
```

Comment:
We could probably simplify this test a bit; see my more minimal repro in https://github.com/python/cpython/issues/112618#issuecomment-1837139058

Reply:
I think that we can keep this test as-is, it is not that complex. But, it also tests that `A("X")` and `A("Y")` is different, which is quite important.

Reply:
I'm specifically not keen on the `get_type_hints()` calls in this test. We don't need them; I don't want this test to start failing if a bug in `get_type_hints()` is introduced; the test should specifically exercise the `Annotated[]` cache.

Either construct instances of `Annotated[]` directly, as in my repro in https://github.com/python/cpython/issues/112618#issuecomment-1837139058, or access it via the `__annotations__` dictionary.